app/src/main/java/pm/gnosis/heimdall/views/StepIndicator.kt

```diff
@@ -0,0 +1,331 @@
+package pm.gnosis.heimdall.views
+
+import android.content.Context
+import android.content.res.ColorStateList
+import android.content.res.TypedArray
+import android.graphics.*
+import android.util.AttributeSet
+import android.util.TypedValue
+import android.view.Gravity
+import android.view.LayoutInflater
+import android.view.View
+import android.widget.LinearLayout
+import android.widget.TextView
+import androidx.core.content.ContextCompat
+import pm.gnosis.heimdall.R
+import kotlin.math.cos
+import kotlin.math.sin
+import kotlinx.android.synthetic.main.view_step_indicator_step.view.*
+import android.graphics.Shader.TileMode
+
+
+class StepIndicator @JvmOverloads constructor(
+    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
+) : LinearLayout(context, attrs, defStyleAttr) {
+
+    private val backgroundPaint: Paint
+
+    init {
+        orientation = HORIZONTAL
+        gravity = Gravity.CENTER
+        backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
+        backgroundPaint.style = Paint.Style.FILL_AND_STROKE
+        backgroundPaint.strokeWidth = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 2f, resources.displayMetrics)
+    }
+
+    fun updateStep(index: Int, stepState: StepState) {
+        val step = getChildAt(index) as Step
+        step.state = stepState
+        invalidate()
+    }
+
+    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
+        if (changed) {
+            var childsMargin = if (childCount == 0) 0 else (width - paddingLeft * 2 - getChildAt(0).measuredWidth * childCount) / childCount
+            var left = l + paddingLeft + childsMargin / 2
+            for (i in 0 until childCount) {
+                val child = getChildAt(i)
+
+                val childWidth = child.measuredWidth
+                val childHeight = child.measuredHeight
+
+                child.layout(left + childsMargin / 2, paddingTop, left + childWidth + childsMargin / 2, childHeight + paddingTop)
+                left += childWidth + childsMargin / 2
+            }
+        }
+    }
+
+    override fun dispatchDraw(canvas: Canvas) {
```

Comment:
what is the difference between `dispatchDraw` and `onDraw`?

Reply:
its called in viewGroup before childs are drawn